echo-pubsub-aws/src/main/java/com/netflix/spinnaker/echo/pubsub/aws/SQSSubscriber.java

```diff
@@ -243,12 +245,44 @@ private static String ensureQueueExists(AmazonSQS amazonSQS,
     return queueUrl;
   }
 
-  // Todo emjburns: pull to kork-aws
+  private void setUpSubscription(AmazonSNS amazonSNS,
+                                 ARN topicARN,
+                                 ARN queueARN,
+                                 AmazonPubsubProperties.AmazonPubsubSubscription subscription){
+    log.debug("Initializing subscription and updating filter policies for subscription {}", subscription.getName());
+    String subscriptionARN = subscribeToTopic(amazonSNS, topicARN, queueARN);
+    applyFilterPolicyToSubscription(amazonSNS, topicARN, queueARN, subscriptionARN, subscription.getFilterPolicy());
+  }
+
   private static String subscribeToTopic(AmazonSNS amazonSNS,
                                          ARN topicARN,
                                          ARN queueARN) {
-    amazonSNS.subscribe(topicARN.getArn(), "sqs", queueARN.getArn());
-    return topicARN.getArn();
+      SubscribeResult subscribeResult = amazonSNS.subscribe(topicARN.getArn(), "sqs", queueARN.getArn());
+      return subscribeResult.getSubscriptionArn();
+  }
+
+  private static void applyFilterPolicyToSubscription(AmazonSNS amazonSNS, ARN topicARN, ARN queueARN, String subscriptionArn, String filterPolicy) {
+    if (filterPolicy == null || filterPolicy.equals("")) {
+      // Refresh subscription to ensure old filterPolicies aren't taking effect.
+      // AWS SNS doesn't have an API for deleting filter policies.
+      refreshSubscription(amazonSNS, topicARN, queueARN, subscriptionArn);
+    } else {
+      try {
+        SetSubscriptionAttributesResult result = amazonSNS.setSubscriptionAttributes(subscriptionArn, "FilterPolicy", filterPolicy);
+      } catch (Exception e) {
+        log.error("Unable to set filter policy for subscription {}", subscriptionArn, e);
+      }
+    }
+  }
+
+  private static String refreshSubscription(AmazonSNS amazonSNS, ARN topicARN, ARN queueARN, String subscriptionArn){
+    log.debug("Refreshing subscription for topic {} and queue {}", topicARN, queueARN);
+    try {
+      amazonSNS.unsubscribe(subscriptionArn);
+    } catch (Exception e) {
+      log.error("Unable to unsubscribe from queue {} from topic {}.", queueARN, topicARN, e);
+    }
+    return subscribeToTopic(amazonSNS, topicARN, queueARN);
```

Comment:
It's a real bummer that AWS forces us into doing something like this.

I could see potential for losing notifications with this pattern if we wind up changing our filters while servicing production notifications. The only work around I can think of off the top of my head... Any time we want to refresh subscriptions, we create a new versioned SQS queue with the updated subscription, unsubscribe the old queue and let it drain, then delete the old SQS queue once we've received empty responses for a few cycles.

My presumption is that Rocket (and other publishers) will be operating on a fire-and-forget design, which leaves us holding the bag on making sure we receive.

Reply:
I feel like our tolerance for missed messages should be extremely low/zero.

I'd be inclined to do as @robzienert has called out ... maintain 1..* queues and monitor draining as needed.